src/install/dracut-install.c

```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
+                size_t len;
+
+                if (!fgets(alias, sizeof(alias), modalias_file))
+                        return 0;
+
+                len = strlen(alias);
+
+                if (len == 0)
+                        return 0;
+
+                if (alias[len - 1] == '\n')
+                        alias[len - 1] = 0;
+
+                return kmod_module_new_from_lookup(ctx, alias, modules);
+        }
+
+        return 0;
+}
+
+static void find_suppliers_for_sys_node(struct kmod_ctx *ctx, Hashmap *suppliers, const char *node_path)
+{
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
+                                if (realpath(dir->d_name, supplier_path)) {
+                                        strncat(supplier_path, "/supplier", sizeof(supplier_path));
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+                                                        hashmap_put(suppliers, supplier_module, supplier_module);
+                                                }
+                                        }
+                                }
+                        }
+                }
+                closedir(d);
+        }
+}
+
+static void find_suppliers_for_module(struct kmod_ctx *ctx, Hashmap *suppliers, const char *module)
+{
+        _cleanup_fts_close_ FTS *fts;
+        char *paths[] = { "/sys/devices/platform", NULL };
+        fts = fts_open(paths, FTS_NOSTAT | FTS_PHYSICAL, NULL);
+
+        for (FTSENT *ftsent = fts_read(fts); ftsent != NULL; ftsent = fts_read(fts)) {
+                if (strcmp(ftsent->fts_name, "modalias") == 0) {
+                        char sysfs_node[PATH_MAX] = {0};
+                        memcpy(sysfs_node, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
+                        sysfs_node[ftsent->fts_parent->fts_pathlen] = '\0';
+
+                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                        struct kmod_list *l = NULL;
+
+                        if (find_module_from_sysfs_node(ctx, sysfs_node, &list) < 0)
+                                continue;
```

Comment:
you keep taking FTS paths, copying them to slice them, then copying them *again* to unslice them.

you can avoid the clumsy copy here and below by getting rid of `sysfs_node` entirely and having
```c
if (find_module_from_sysfs_node(ctx, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen, &list) < 0)
```
coupled with
```c
static int find_module_from_sysfs_node(struct kmod_ctx *ctx, const char *sysfs_node, int sysfs_node_len, struct kmod_list **modules)
{
        char modalias_path[PATH_MAX] = {0};
        snprintf(modalias_path, sizeof(modalias_path), "%.*s/modalias", sysfs_node_len, sysfs_node);
        _cleanup_fclose_ FILE *modalias_file = NULL;
```
and similarly `find_suppliers_for_sys_node()` as
```c
find_suppliers_for_sys_node(ctx, suppliers, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
```
and
```c
find_suppliers_for_sys_node(ctx, suppliers, parent->fts_path, parent->fts_pathlen);
```
and have *that* do
```c
static void find_suppliers_for_sys_node(struct kmod_ctx *ctx, Hashmap *suppliers, const char *node_path_raw, size_t node_path_len)
{
        char node_path[PATH_MAX] = {};
        memcpy(node_path, node_path_raw, node_path_len);
        DIR *d = opendir(node_path);
        struct dirent *dir;
```